### ISSUE TITLE
GPKG: creation of spatial index is quite slow

### ISSUE THREAD

**Issue by theroggy** [OSGeo/gdal#7614]
Not really a new observation, but the creation of a spatial index on a (larger) Geopackage file is very slow. I made a [test script](https://github.com/OSGeo/gdal/files/11261889/create_index_gpkg_perf.txt) in python to illustrate this.

It creates a grid with 1.002.001 squares, writes this to .shp and .gpkg and creates spatial indexes on them. The output:
```
writing test.shp without spatial index took 0:00:14.067565
create spatial index on test.shp took 0:00:04.890439
writing test.gpkg without spatial index took 0:00:06.265831
writing test_si.gpkg with spatial index took 0:00:38.086723
create spatial index on test.gpkg with default sqlite cache took 0:00:36.834020
create spatial index on test.gpkg with 512 MB sqlite cache took 0:00:37.406399
```

I wonder, is it clear why this is slow or where the bottleneck is located? In GDAL, Spatialite, SqLite, ???

Where would be a logical place to post this issue to try to get this improved? Or have such efforts already been tried?

**Comment by theroggy** [comment]
I was able to reproduce without any gdal or spatialite involved that the time being taken to create the index is indeed ~100% the "INSERT INTO my_tree...", so 100% sqlite.

I had another look at the qix format, and as far as I can tell it is a slightly different tree implementation (quadtree)... so I'll have a look if I can get some timings using e.g. https://rtree.readthedocs.io/en/latest/ so the time comparison is as "comparable" as possible.

**Comment by jratike80** [comment]
The R*Tree module https://www.sqlite.org/rtree.html is probably more complicated than it may feel. Maybe the fastest possible writing has not been the main design goal. Normally indexes are read much more often that written, but naturally a data provider has a different point of view.
I wonder if it could be faster to create simple indexed minx/maxx/miny/maxy columns and how much slower it would be use those in bbox queries.

**Comment by rouault** [comment]
> I wonder if it could be faster to create simple indexed minx/maxx/miny/maxy columns and how much slower it would be use those in bbox queries.

Cf the NGA GeoInt Geometry Index extension: http://ngageoint.github.io/GeoPackage/docs/extensions/geometry-index.html . It probably needs also additional creation of regular index on the min_x, max_x, min_y, max_y columns

**Comment by rouault** [comment]
> It probably needs also additional creation of regular index on the min_x, max_x, min_y, max_y columns

actually no, that's detrimental to performance in some cases.
And the performance boost of using nga_geometry_index is "just" a x1.6 over using the bounding box included in the GeoPackage geometry blob (with https://github.com/OSGeo/gdal/pull/7621 applied):
```
$ time ogrinfo nz-building-outlines-no-index.gpkg -sql "select count(*) from \"nz-building-outlines\" where ST_EnvelopesIntersects(geometry, 1500000.0,5200000.0,1800000.0,5800000.0)" -al -q

Layer name: SELECT
OGRFeature(SELECT):0
  count(*) (Integer) = 701300


real	0m0,563s
user	0m0,294s
sys	0m0,267s

$ time ogrinfo nz-building-outlines-ngageoint.gpkg -sql "select count(*) from nga_geometry_index where max_x>=1500000 and max_y>= 5200000 and min_x <=1800000 and min_y <= 5800000" -al -q

Layer name: SELECT
OGRFeature(SELECT):0
  count(*) (Integer) = 701300


real	0m0,360s
user	0m0,290s
sys	0m0,068s
```

**Comment by jratike80** [comment]
It would be nice to see also the timing for R*Tree.

**Comment by rouault** [comment]
> It would be nice to see also the timing for R*Tree.
```
$ time ogrinfo nz-building-outlines.gpkg -sql "SELECT count(*) FROM \"rtree_nz-building-outlines_geometry\" WHERE maxx >= 1500000 AND minx <= 1800000 AND maxy >= 5200000 AND miny <= 5800000" -al -q

Layer name: SELECT
OGRFeature(SELECT):0
  count(*) (Integer) = 701301


real	0m0,146s
user	0m0,120s
sys	0m0,024s
```
And the RTree is magnitude order much faster than linear scanning when selecting a very small number of objects (here this selects ~ 20% of the table)

**Comment by jratike80** [comment]
So it feels like the creation of the R*Tree index is slower for a reason and it gets paid back rather soon when the data are queried.

**Comment by theroggy** [comment]
As mentioned above I had a look to find an rtree implementation to benchmark against because .qix uses a quandtree. I don't know any details on the difference, but comparing the same algoritm should be a fairer comparison.

In geopandas one of the available indexes is an rtree, so I had a look at that. Apparently the [rtree](https://github.com/Toblerity/rtree0) library used is a python wrapper around a c library. Not sure how fast or slow it is... but for the sample data above creating the index takes ~9 seconds compared to the 38 seconds to fill up the index in sqlite.

The 9 s is without serializing to file, but still it gives the impression there might be room for improvement in the sqlite implementation.

An important thing to note is that the "rtree" library has an explicit [bulk/stream load](https://rtree.readthedocs.io/en/latest/performance.html#use-stream-loading) option that would be a lot faster. I tested both, and the 9 seconds is for the bulk load way, filling the index in a row-by-row loop way took over 2 minutes... Not sure if python overhead in the loop plays a significant role or not, but that way it is a lot slower than sqlite.

Possibly such a bulk load option might be the trick that could boost the initial spatial index creation in sqlite as well?

Code snippet of my rtree test:
```
import datetime
from rtree import index

# Prepare test data
x, y = (0, 0)
rects = []
size = 10
while y <= 10000:
    while x <= 10000:
        rects.append((x, y, x + size, y + size))
        x += size
    x = 0
    y += size

# Create rtree
start = datetime.datetime.now()
data = ((i, item, None) for i, item in enumerate(rects))
index_properties = index.Property(type=index.RT_RTree)
idx = index.Index(data, properties=index_properties)
print(f"create rtree took {datetime.datetime.now() - start}")
```

**Comment by jratike80** [comment]
I fear that for making the creation of the R\*Tree index in SQLite faster one should have a thorough understanding about the virtual table mechanism of SQLite https://www.sqlite.org/vtab.html, and what the R\*Tree module is doing when it creates the contents to the shadow tables ...node, ...parent, and ...rowid.
I suggest to write mail to SQLite developers and ask about the design principles of the R*Tree module.

There may well be alternative methods for writing a spatial index faster. I guess that such system would require a vendor specific extension because the GeoPackage standard requires that the gpkg_rtree_index extension is using the standard module

> Requirements
> This extension uses the rtree implementation provided by the SQLite R*Tree Module extension
> documented at http://www.sqlite.org/rtree.html.
>

**Comment by jratike80** [comment]
Hey @rouault , what if you continued what you did in https://github.com/OSGeo/gdal/pull/7621, and fool the standard R*Tree to index the bounding boxes inside the GeoPackage BLOBs instead of the real geometries?

**Comment by theroggy** [comment]
> I suggest to write mail to SQLite developers and ask about the design principles of the R*Tree module.

It is indeed the plan to post in the SQLite support forum and ask if they see an option to speed up the creation of the index.

> ... and fool the standard R*Tree to index the bounding boxes inside the GeoPackage BLOBs instead of the real geometries? Or is it done already? 

The R*Tree index is already populated using the bounding boxes. 

> Where do the ST_MinX etc. functions find the values?

The GeoPackage BLOB (typically) already contains the bounding box, so ST_MinX etc. can just read them directly without any further processing. Additionally I tested a few days ago if it made a difference to first copy the ST_MinX,... of the geometries to a simple table, and then create the rtree index based on that, but this didn't give any difference, so hence my conclusion that ~100% of the time is spent building/inserting into the rtree index.

**Comment by rouault** [comment]
> and fool the standard R*Tree to index the bounding boxes inside the GeoPackage BLOBs instead of the real geometries? Or is it done already?

yes, that's already done. In https://github.com/OSGeo/gdal/blob/master/ogr/ogrsf_frmts/gpkg/ogrgeopackagetablelayer.cpp#L2427, in CreateFeature(), OGR collects the bounding box computed to create the  header GeoPackage geometry blob, store that in a memory queue, builds a RTree in a dummy in-memory sqlite database in a thread, and then copies the shadow tables to the real database at the end. That was implemented per https://github.com/OSGeo/gdal/pull/6573. There's really nothing more we can done on OGR side, without going into really low level details of the SQLite R*Tree implementation. That is messing directly with the content of the shadow tables (which, as far as I can see, isn't documented in plain language). Non trivial effort, and it is not obvious if we could beat SQLite implementation (perhaps...). It would be better if SQLite itself could have a fast bulk mode implemented, if that's ever possible given its architecture

**Comment by theroggy** [comment]
Hopefully there is still a bigger speedup possible, but I had an interesting find.

I'm using python via conda(-forge) on windows for my tests and timings. When running testing my technology indendent sql script with a different sqlite.exe I noticed that this was significantly faster: 18 secs instead of 31 secs.

So, I already posted a issue for that aspect here: https://github.com/conda-forge/sqlite-feedstock/issues/89
